tests/cases/compiler/correlatedUnions.ts

```diff
@@ -301,3 +301,76 @@ function getValueConcrete<K extends keyof Foo1>(
 ): Foo1[K] | undefined {
   return o[k];
 }
+
+// repro from https://github.com/microsoft/TypeScript/issues/54680

```

Comment:
So this regression was assessed as working as intended by @ahejlsberg [here](https://github.com/microsoft/TypeScript/issues/54680) but arguably the same unsoundness can be observed in different scenarios (see [here](https://github.com/microsoft/TypeScript/issues/54824)). 

The new test case from https://github.com/microsoft/TypeScript/issues/54834 doesn't manifest the same unsoundness from what I can tell though.